cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.pipeline.kafka;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import co.cask.cdap.logging.appender.kafka.LoggingEventSerializer;
+import co.cask.cdap.logging.meta.Checkpoint;
+import com.google.common.collect.ImmutableMap;
+import kafka.api.FetchRequest;
+import kafka.api.FetchRequestBuilder;
+import kafka.api.PartitionOffsetRequestInfo;
+import kafka.common.ErrorMapping;
+import kafka.common.TopicAndPartition;
+import kafka.javaapi.FetchResponse;
+import kafka.javaapi.OffsetRequest;
+import kafka.javaapi.OffsetResponse;
+import kafka.javaapi.consumer.SimpleConsumer;
+import kafka.message.MessageAndOffset;
+import org.apache.twill.kafka.client.BrokerInfo;
+import org.apache.twill.kafka.client.BrokerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Resolve matching Kafka offset with given checkpoint.
+ */
+public class KafkaOffsetResolver {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaOffsetResolver.class);
+
+  // TODO: (CDAP-8439) determine the appropriate size
+  private static final int FETCH_SIZE = 1024 * 1024;        // Use a default fetch size.
+  private static final int SO_TIMEOUT = 5 * 1000;           // 5 seconds.
+  private static final long LATEST = kafka.api.OffsetRequest.LatestTime();
+  private static final long EARLIEST = kafka.api.OffsetRequest.EarliestTime();
+  // A rough estimation of Kafka publishing delay of a message from the original cluster and the destination cluster
+  // during cross-cluster replication
+  private static final long REPLICATION_DELAY = 5 * 60 * 1000;
+  // The max mount of time that the log event time of a message with smaller offset
+  // can exceed a message with larger offset
+  private static final long MAX_TIME_GAP = 60 * 1000; // 1 min
```

Comment:
This time seems not equivalent to `KafkaPipelineConfig.eventDelayMillis`. `KafkaPipelineConfig.eventDelayMillis` is the maximum range of time within which logs are guaranteed to be ordered by time, but `MAX_TIME_GAP` is the maximum amount of time that messages can come in a wrong order